compiler/rustc_expand/src/mbe/metavar_expr.rs

```diff
@@ -0,0 +1,159 @@
+use std::str::FromStr;
+
+use rustc_ast::token::{self, Lit};
+use rustc_ast::tokenstream::{Cursor, TokenStream, TokenTree};
+use rustc_ast_pretty::pprust;
+use rustc_errors::{Applicability, PResult};
+use rustc_session::parse::ParseSess;
+
+use rustc_span::symbol::Ident;
+use rustc_span::Span;
+
+/// A meta-variable expression, for expansions based on properties of meta-variables.
+#[derive(Debug, Clone, PartialEq, Encodable, Decodable)]
+crate enum MetaVarExpr {
+    /// The number of repetitions of an identifier, optionally limited to a number
+    /// of outer-most repetition depths. If the depth limit is `None` then the depth is unlimited.
+    Count(Ident, Option<usize>),
+
+    /// Ignore a meta-variable for repetition without expansion.
+    Ignore(Ident),
+
+    /// The index of the repetition at a particular depth, where 0 is the inner-most
+    /// repetition. The `usize` is the depth.
+    Index(usize),
+
+    /// The length of the repetition at a particular depth, where 0 is the inner-most
+    /// repetition. The `usize` is the depth.
+    Length(usize),
+}
+
+impl MetaVarExpr {
+    /// Attempt to parse a meta-variable expression from a token stream.
+    crate fn parse<'sess>(
+        input: &TokenStream,
+        sess: &'sess ParseSess,
+    ) -> PResult<'sess, MetaVarExpr> {
+        let mut tts = input.trees();
+        match tts.next() {
+            Some(TokenTree::Token(token)) if let Some((ident, false)) = token.ident() => {
+                let Some(TokenTree::Delimited(_, token::Paren, args)) = tts.next() else {
+                    let msg = "meta-variable expression parameter must be wrapped in parentheses";
+                    return Err(sess.span_diagnostic.struct_span_err(ident.span, msg));
+                };
+                let mut iter = args.trees();
+                let rslt = match &*ident.as_str() {
+                    "count" => parse_count(&mut iter, sess, ident.span)?,
+                    "ignore" => MetaVarExpr::Ignore(parse_ident(&mut iter, sess, ident.span)?),
+                    "index" => MetaVarExpr::Index(parse_depth(&mut iter, sess, ident.span)?),
+                    "length" => MetaVarExpr::Length(parse_depth(&mut iter, sess, ident.span)?),
+                    _ => {
+                        let err_msg = "unrecognized meta-variable expression";
+                        let mut err = sess.span_diagnostic.struct_span_err(ident.span, err_msg);
+                        err.span_suggestion(
+                            token.span,
+                            "supported expressions are count, ignore, index and length",
+                            String::new(),
+                            Applicability::MachineApplicable,
+                        );
+                        return Err(err);
+                    }
+                };
+                Ok(rslt)
+            }
+            _ => return Err(sess.span_diagnostic.struct_err("expected meta-variable expression"))
```

Comment:
No test case for this error.

Reply:
Also, `parse_iter` should probably be reused for this function, it's exactly the same case as inside `count`/`ignore`.